File: src/main/java/com/woomulwoomul/woomulwoomulbackend/common/log/InterceptorLogging.java
package com.woomulwoomul.woomulwoomulbackend.common.log;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.springframework.lang.Nullable;
import org.springframework.util.StopWatch;
import org.springframework.web.servlet.HandlerInterceptor;
import org.springframework.web.servlet.ModelAndView;

import java.util.UUID;

@Slf4j
@Getter
public class InterceptorLogging implements HandlerInterceptor {

    private static final ThreadLocal<StopWatch> stopWatchThreadLocal = new ThreadLocal<>();
    private static final ThreadLocal<String> uuidThreadLocal = new ThreadLocal<>();

    public boolean preHandle(HttpServletRequest request,
                             @Nullable HttpServletResponse response,
                             @Nullable Object handler) {
        String uuid = UUID.randomUUID().toString();
        String method = request.getMethod();
        String uri = request.getRequestURI();

        StopWatch stopWatch = new StopWatch();
        stopWatch.start();

        log.info("========== [{} | START] {} {} ==========", uuid, method, uri);

        uuidThreadLocal.set(uuid);
        stopWatchThreadLocal.set(stopWatch);

        return true;
    }

    public void postHandle(HttpServletRequest request,
                           @Nullable HttpServletResponse response,
                           @Nullable Object handler,
                           ModelAndView modelAndView) {
        StopWatch stopWatch = stopWatchThreadLocal.get();
        stopWatch.stop();

        long time = stopWatch.getTotalTimeMillis();
        String uuid = uuidThreadLocal.get();
        String method = request.getMethod();
        String uri = request.getRequestURI();

        log.info("========== [{} | FINISH] {} {} | time={}ms ==========", uuid, method, uri, time);

        uuidThreadLocal.remove();
        stopWatchThreadLocal.remove();
    }

    public static String getRequestId() {
        return uuidThreadLocal.get();
    }
}

File: core/src/main/kotlin/com/woomulwoomul/core/common/log/InterceptorLogging.kt
package com.woomulwoomul.core.common.log

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.LoggerFactory
import org.springframework.util.StopWatch
import org.springframework.web.servlet.HandlerInterceptor
import org.springframework.web.servlet.ModelAndView
import java.util.*

class InterceptorLogging : HandlerInterceptor {

    companion object {
        private val log = LoggerFactory.getLogger(this::class.java)
    }

    private val stopWatchThreadLocal: ThreadLocal<StopWatch> = ThreadLocal()
    private val uuidThreadLocal: ThreadLocal<String> = ThreadLocal()

    val requestId:String = uuidThreadLocal.toString().substringAfterLast("@")

    override fun preHandle(request: HttpServletRequest, response: HttpServletResponse, handler: Any): Boolean {
        val uuid = UUID.randomUUID().toString()
        val method = request.method
        val uri = request.requestURI
        val stopWatch = StopWatch()

        stopWatch.start()

        log.info("========== [{} | START] {} {} ==========", uuid, method, uri)

        uuidThreadLocal.set(uuid)
        stopWatchThreadLocal.set(stopWatch)

        return super.preHandle(request, response, handler)
    }

    override fun postHandle(
        request: HttpServletRequest,
        response: HttpServletResponse,
        handler: Any,
        modelAndView: ModelAndView?
    ) {
        val stopWatch = stopWatchThreadLocal.get()

        stopWatch.stop()

        val time = stopWatch.totalTimeMillis
        val uuid = uuidThreadLocal.get()
        val method = request.method
        val uri = request.requestURI

        log.info("========== [{} | FINISH] {} {} | time={}ms ==========", uuid, method, uri, time)

        uuidThreadLocal.remove()
        stopWatchThreadLocal.remove()
    }
}